### ISSUE TITLE
Problem with UART :  UART event queue full

### ISSUE THREAD

**Issue by Liberasys** [loboris/MicroPython_ESP32_psRAM_LoBo#50]
Hello,
I am trying to use UART2. For this, I try this code:
`import machine
import time
uart = machine.UART(2, tx=16, rx=17)
uart.init(115200, bits=8, parity=None, stop=1, timeout=10000)
uart.readline()`

After having run this code, I print "hello" three times with a connected software (putty on /dev/ttyS0, connected to a CP2102 USB to UART Bridge Controller, connected to TX2 and RX2 Pins).

This is what I get on the REPL:
`MicroPython ESP32_LoBo_v3.1.14 - 2017-01-29 on ESP32 board with ESP32
Type "help()" for more information.
>>> import machine
>>> import time
>>> uart = machine.UART(2, tx=16, rx=17)
>>> uart.init(115200, bits=8, parity=None, stop=1, timeout=10000)
>>> uart.readline()
W (14588) uart: UART event queue full
W (14692) uart: UART event queue full
W (15597) uart: UART event queue full
W (15821) uart: UART event queue full
W (15989) uart: UART event queue full
W (16125) uart: UART event queue full
W (16293) uart: UART event queue full
W (16453) uart: UART event queue full
Task watchdog got triggered. The following tasks did not reset the watchdog in time:
 - mp_task (CPU 0/1)
Tasks currently running:
CPU 0: Telnet
CPU 1: IDLE
b'hello\rhello\rhello\r'
>>>
`

I think I am missing something or maybe there is a problem in the firmware because:
- readline() doesn't seems to work as expected (returns the full buffer instead of returning only one line)
- UART event queue doesn't seems to be consumed.

Please help Boris ?
Thank you, Gautier.

**Comment by elektrobil** [comment]
Hi Lobo and  Liberasys 
I have same problem too with UART its becouse I dont writre a new issue. Im reading some data with UART but ture data returns only 21 times (21x12=252byte) after 66th cycle returns true data again until 88th cycle
`var = uart.read(12)`
How is that possible and Is this the source of the error because of the code I wrote? 
P.S. : This code running in a thread

King Regards

**Comment by loboris** [comment]
Thanks for reporting this.
**machine.UART** is now updated, bug fixed and should function as intended.

One new argument is added to the `machine.UART.init()`: `buffer_size`
The default buffer size is 256 bytes and can be now set in the range `256` - `4096`
For example:
`uart = machine.UART(2, tx=22, rx=21, buffer_size=1024)` 

The full documentation for the UART module will be added soon.

**Comment by elektrobil** [comment]
Thank you for your update @loboris  but still same error how can I clean the buffer every cycle
`true data = b'2.012/022\x00\r\n'`
after fill the buffer 98th cycle 
`wrong data = b'2/022\x00\r\n3.26'`

I want to clean buffer every cycle in my loop

Thank you so much for your help
King regards

**Comment by loboris** [comment]
@elektrobil
I'm not shure if I understand the problem.
Maybe you could give more detailes, what method you are using to read, what the other side sends, ...

If you need to clean the input buffer, you should read until nothing is returned.

I'm planning to update the UART module with more methods like input buffer clear, read with timeout, read until a specified pathern, etc.

**Comment by elektrobil** [comment]
Hi @loboris thank you for your helps
My project is nrf24le1 reads the sensor and send to the main nrf24le1. Main Nrf chip connected to Esp32 via Uart.

`import machine `

`uart = machine.UART(2, tx=pin1, rx=pin2, buffer_size=1024) `
`uart.init(9600, bits=8, parity=None, stop=1, timeout=10000)`
`var = uart.read(12)`

This code runs in a thread function continious
Everything ok untill buffer size full 
true data = b'2.012/022\x00\r\n'
after fill the buffer 98th cycle (buffer size 1024)(when buffersize =256 cycle return error at 21st cycle)
wrong data = b'2/022\x00\r\n3.26' last 3.26 must be in front of the variable.
I think I need to clean input buffer. Is that possible with micropython code because Im new in micropython
Thank you so much again

**Comment by loboris** [comment]
Could you please post your complete thread function. I think there should be a simple solution for that.

**Comment by elektrobil** [comment]
Hi @loboris this is my thread function below Thank you for your help
`       gc.collect()`
`        i=0`
`        inx=0`
`        htext = 35`
`        vtext = 10`
`        if uart.any():`
`            gelen = uart.read(12)`received data = b'2.012/022\x00\r\n
`            l+=1`
`            print(gelen)`
`            print("SAYI = ",l)#count loop cycle`
`            gelens = gelen.decode()`
`            print(len(gelens))`
`            print(gelens)`
`            gelens = gelens.rstrip('\n')`
`            gelens = gelens.rstrip('\r')`
`            gelens = gelens.rstrip('\x00')`
`            if len(gelens) == 9:`
`                gelens,sets = gelens.split('/')`
`                sb = list(sets)`
`                print(sb)`
`                if int(sb[1])>1:`
`                    inx = int(sb[2]) + (4 * (int(sb[1])-1))`
`                else:`
`                    inx = int(sb[2])`
`                vals[inx-1] = float(gelens)`
King Regards...

**Comment by loboris** [comment]
I suppose this code is included in an infinite loop in your thread function ?
If not, it will run only once.

`uart.read(12)` reads 12 bytes from uart buffer and removes them from the buffers, so if it is executed regularly no buffer overflow should occur.
 
You can use `if uart.any() >= 12:` to test if at leas 12 bytes have arrived.

The UART module is updated today, some new functions are added as well as callbacks on some uart events. Full documentation will be on Wiki tomorrow.

Note:
When you insert the code here, paste the code, select all code and press **<>**.
The code should apear betveen two sets of **```**
That way the code structure will be preserved.

**Comment by elektrobil** [comment]
I think buffer can't clean and I have the same error. How could I be sure to clean the buffer ?
This sample : using `uart = machine.UART(2, tx=pin1, rx=pin2, buffer_size=256)`
(it seems to me) Buffer Not Full Output:
**b'3.269/023\x00\r\n'**
**SAYI =  21**
12 (chars)
3.269/023 -- (mydata/id)
['0', '2', '3']
-------------------------------------- after 21st cycle it means 21x12 = 252bytes
(it seems to me)Buffer Full Output:
**b'2/022\x00\r\n3.26'**
**SAYI =  22**
12
2/022
3.26

Thank you for help on saturday night

**Comment by elektrobil** [comment]
Hi @loboris  I pulled the new version but data comes like that.
```
0;0.0;0.0;0.
b'0;0.0;0.0;0.'
SAYI =  3
```

**Comment by elektrobil** [comment]
Thank you @loboris 
UART works good 
King Regards

**Comment by loboris** [comment]
Please check the new [UART Wiki page](https://github.com/loboris/MicroPython_ESP32_psRAM_LoBo/wiki/uart).
There are some new methods, and you can now use callback in your code.
For your application which reads the messages ending with '\r\n' you can use the new method `uart.readln()`.

**Comment by elektrobil** [comment]
I sow Wiki page when you commit. Thank you so much. It works perfect

**Comment by Liberasys** [comment]
Hello, UART update seems great Boris ! UART callback with pattern is very cool !
I tested with putty as serial terminal client under Linux. uart.readln() works great with lineend='\r'.
Thank you Boris !
